### DISCUSSION TITLE
Issue w/ random Google Acct. Logouts

### DISCUSSION THREAD

**Discussion by HIT2022** [leikoilja/ha-google-home#126] [Q&A] [answered]
Hi! `ha-google-home` has been working great for me for the last week or so... but I noticed that I've been randomly getting logged out of all devices (Android, Google TV, etc.). Not sure if it has something to do with the frequency that this component is pulling down tokens/updates, but each device will prompt me to re-login once every 3 days or so. Has anyone else been running into this, and/or any suggestions on how I might address it?

**Comment by leikoilja** [comment]
hmmm, if i recall correctly, @DurgNomis-drol had something similar at the very beginning. 

@HITChris, are you using the latest integration version? 
We have had that in mind when working on the integration and there are few mechanisms in place which are taking into consideration to **not** generate any new tokens unless they are needed. The requests made to google devices to fetch alarms/tokens should not play any role at those logout issues, thou I'm wondering if there are any other security mechanisms from Google that might be triggered 🤔

**Reply by Drakulix** [reply]
I have had the same issue previously, when working with the unofficial google music api. It seems to be caused by the authentication flow, which mimics the android login. The only way to fix this, was to cache the generated master token instead of always performing a login using email and password.

Is there any chance ha-google-home might not be caching the master_token correctly or that glocaltokens is doing a full login anyway? Maybe some log messages could reveal what happens here?

**Comment by DurgNomis-drol** [comment, marked as answer]
@leikoilja I still experience this if i reboot my HA instance a few times in a row. But not always. 

As @Drakulix says, it happens if you try to retrieve the master token to often. Which we should have accounted for.

**Reply by KapJI** [reply]
I think we should store master token in `ConfigEntry`. Don't we do this yet?

**Reply by DurgNomis-drol** [reply]
Yes we do. That is what is confusing me a bit.

**Reply by DurgNomis-drol** [reply]
I just looked in the code, and we save the master token, but we don't use it after a restart. I am on my phone but the code is in `api.py` when we initiate `glocaltokensclient`. But as far as I know, `glocaltokens` supports passing master token.

**Reply by leikoilja** [reply]
Ops, that's probably it, good find. Fixing this issue we can possibly help to eliminate random logouts a bit. @DurgNomis-drol, can you please create an issue for that so we could take it from there? 👍

**Reply by DurgNomis-drol** [reply]
#130

**Reply by DurgNomis-drol** [reply]
Another thing we should probably introduce down the road, is reauth if the user changes his password.

**Reply by HIT2022** [reply]
#130 seems to have resolved my issue. Thanks All!
